pyqtorch/noise/protocol.py

```diff
@@ -12,43 +14,144 @@ class NoiseProtocol:
-        if self.target:
-            return (
-                f"{self.protocol}(prob: {self.error_probability}, "
-                f"target: {self.target})"
-            )
-        return f"{self.protocol}(prob: {self.error_probability})"
+class NoiseProtocol:
+    """
+    Defines a protocol of noisy quantum channels.
+
+    Args:
+        protocol: single NoiseType instance or list of NoiseType instances, or
+            list of (NoiseType, options) tuple. When passing list of tuples, for
+            each NoiseType the options should be a dict containing the "error_probability",
+            and optionally a "target". If no "target" is present, the noise instance
+            will be applied to the same target of the gate it is used on.
+        error_probability: probability of error when passing a single NoiseType
+            or a list of NoiseTypes. Note that all noise types require a single float for the
+            error_probability, while the PauliChannel and GeneralizedAmplitudeDamping require
+            a tuple of error probabilities.
+
+    Examples:
+    ```
+        from pyqtorch.noise import NoiseProtocol, NoiseType
+
+        # Single noise instance
+        protocol = NoiseProtocol(NoiseType.BITFLIP, error_probability = 0.5)
+        protocol = NoiseProtocol.bitflip(error_probability = 0.5)
+
+        # Multiples noise instances with same probability
+        protocol = NoiseProtocol(
+                        [NoiseType.BITFLIP, NoiseType.PHASEFLIP],
+                        error_probability = 0.5
+                    )
+
+        # Multiples noise instances with different options
+        prot0 = NoiseProtocol.bitflip(0.5)
+        prot1 = NoiseProtocol.pauli_channel((0.1, 0.2, 0.7))
+        protocol = NoiseProtocol([prot0, prot1])
+
+    ```
+    """
+
+    def __init__(
+        self,
+        protocol: NoiseType | list[NoiseType] | list[NoiseProtocol],
+        error_probability: tuple[float, ...] | float | None = None,
+        target: int | None = None,
+    ) -> None:
+
+        self._error_probability = error_probability
+        self._target = target
+
+        if isinstance(protocol, NoiseType):
+            self.noise_instances = [(protocol, error_probability, target)]
+        elif isinstance(protocol, list) and isinstance(protocol[0], NoiseType):
+            self.noise_instances = [(p, error_probability, target) for p in protocol]  # type: ignore [misc]
+        elif isinstance(protocol, list) and isinstance(protocol[0], NoiseProtocol):
+            self.noise_instances = []
+            for p in protocol:
+                self.noise_instances.extend(p.noise_instances)  # type: ignore [union-attr]
+        else:
+            raise TypeError("Incorrect protocol argument.")
+
+        for _, error_probability, _ in self.noise_instances:
+            if error_probability is None:
+                raise ValueError(
+                    "This protocol requires a value to the error_probability argument"
```

Comment:
```suggestion
                    f"Protocol {protocol} requires a non-negative error_probability argument of type `float`. Got {error_probability}."
```

Actually, checking that error probabilities are non-negative might be a good check to add here.

Reply:
So the test here is a bit different: currently it is just meant to check if the user forgot to pass an `error_probability` (since it is an optional argument in the `__init__`, but only when calling `NoiseProtocol([prot1, prot2])`). Currently the test for the actual contents of the error probabilities is down downstream in the instantiation of the actual noise gates, since a few of them do a different check.